src/vstest.console/CommandLine/CommandLineOptions.cs

```diff
@@ -283,28 +284,45 @@ public void AddSource(string source)
             {
                 throw new CommandLineException(CommandLineResources.CannotBeNullOrEmpty);
             }
-
+            
             source = source.Trim();
-
+            
             // Convert the relative path to absolute path
-            if(!Path.IsPathRooted(source))
+            if (!Path.IsPathRooted(source))
             {
                 source = Path.Combine(FileHelper.GetCurrentDirectory(), source);
             }
 
-            if (!FileHelper.Exists(source))
-            {
-                throw new CommandLineException(
-                    string.Format(CultureInfo.CurrentUICulture, CommandLineResources.TestSourceFileNotFound, source));
-            }
+            var filePatternParser = new FilePatternParser();
+            var sourceFiles = new List<string>();
+            var isValidPattern = filePatternParser.IsValidPattern(source, out sourceFiles);
 
-            if (this.sources.Contains(source, StringComparer.OrdinalIgnoreCase))
+            // If the given file is a full path and not a pattern
+            if (!isValidPattern)
```

Comment:
combine these two, and the have the for loop first and return.

Have the code for file exist outside the if block, the return above would make this block serve like an else block

Reply:
Done.